File: retrocpu_asm/src/main/cpuType.ts
import { stripLineComment } from "./parser";
import type { CpuType } from "./types";

/** `.cpu` / `--cpu` で指定できる CPU */
export const CPU_TYPES: readonly CpuType[] = ["mn1613", "tms9995"];

/**
 * CPU 名を正規化する（大文字小文字無視）。
 * @param value 入力文字列
 * @returns 既知なら CpuType。不明なら null
 */
export function parseCpuType(value: string | undefined): CpuType | null {
  if (value === undefined) return null;
  const v = value.trim().toLowerCase();
  if (v === "mn1613" || v === "tms9995") return v;
  return null;
}

/**
 * 行末コメントを除いた本文を返す。
 * @param line 1 行
 * @returns トリム済み本文
 */
function lineBody(line: string): string {
  return stripLineComment(line).trim();
}

/**
 * ソース先頭（コメント・空行を除く）の `.cpu` を読む。
 * `.cpu` が後段にある、値が不正、重複、オペランド無しはエラー。
 * @param sourceText アセンブラソース全文
 * @returns `.cpu` があればその CPU。無ければ undefined
 */
export function scanSourceCpu(sourceText: string): CpuType | undefined {
  const lines = sourceText.replace(/\r\n/g, "\n").split("\n");
  let seenContent = false;
  let found: CpuType | undefined;

  for (let i = 0; i < lines.length; i += 1) {
    const body = lineBody(lines[i]!);
    if (!body) continue;

    const m = body.match(/^\.cpu(?:\s+(\S+))?$/i);
    if (m) {
      if (seenContent) {
        throw new Error(
          `Line ${i + 1}: .cpu must be the first non-comment line`,
        );
      }
      if (found !== undefined) {
        throw new Error(`Line ${i + 1}: duplicate .cpu directive`);
      }
      const raw = m[1];
      if (!raw) {
        throw new Error(`Line ${i + 1}: .cpu requires mn1613 or tms9995`);
      }
      const cpu = parseCpuType(raw);
      if (!cpu) {
        throw new Error(
          `Line ${i + 1}: unknown .cpu '${raw}' (mn1613 / tms9995)`,
        );
      }
      found = cpu;
      seenContent = true;
      continue;
    }

    seenContent = true;
  }

  return found;
}

/**
 * CLI / 引数の CPU とソース `.cpu` から実際に使う CPU を決める。
 * 引数があればそれを優先。どちらも無ければエラー。
 * @param explicit `--cpu` / `assemble()` 第 2 引数。未指定は undefined
 * @param sourceText アセンブラソース全文
 * @returns 確定した CPU
 */
export function resolveCpuType(
  explicit: CpuType | undefined,
  sourceText: string,
): CpuType {
  const fromSource = scanSourceCpu(sourceText);
  if (explicit) return explicit;
  if (fromSource) return fromSource;
  throw new Error(
    "CPU が未指定です（--cpu / -m または先頭の .cpu で mn1613 / tms9995 を指定してください）",
  );
}

File: retrocpu_asm_rs/src/cpu_type.rs
use crate::error::AsmError;
use crate::parser::strip_line_comment;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CpuType {
    Mn1613,
    Tms9995,
}

pub fn parse_cpu_type(value: Option<&str>) -> Option<CpuType> {
    let v = value?.trim().to_ascii_lowercase();
    match v.as_str() {
        "mn1613" => Some(CpuType::Mn1613),
        "tms9995" => Some(CpuType::Tms9995),
        _ => None,
    }
}

pub fn scan_source_cpu(source_text: &str) -> Result<Option<CpuType>, AsmError> {
    let mut seen_content = false;
    let mut found: Option<CpuType> = None;

    for (idx, line) in source_text.replace("\r\n", "\n").split('\n').enumerate() {
        let body = strip_line_comment(line).trim().to_string();
        if body.is_empty() {
            continue;
        }

        let parts: Vec<&str> = body.split_whitespace().collect();
        if !parts.is_empty() && parts[0].eq_ignore_ascii_case(".cpu") {
            if seen_content {
                return Err(AsmError::new(format!(
                    "Line {}: .cpu must be the first non-comment line",
                    idx + 1
                )));
            }
            if found.is_some() {
                return Err(AsmError::new(format!(
                    "Line {}: duplicate .cpu directive",
                    idx + 1
                )));
            }
            if parts.len() < 2 {
                return Err(AsmError::new(format!(
                    "Line {}: .cpu requires mn1613 or tms9995",
                    idx + 1
                )));
            }
            let cpu = parse_cpu_type(Some(parts[1])).ok_or_else(|| {
                AsmError::new(format!(
                    "Line {}: unknown .cpu '{}' (mn1613 / tms9995)",
                    idx + 1,
                    parts[1]
                ))
            })?;
            found = Some(cpu);
            seen_content = true;
            continue;
        }

        seen_content = true;
    }

    Ok(found)
}

pub fn resolve_cpu_type(explicit: Option<CpuType>, source_text: &str) -> Result<CpuType, AsmError> {
    if let Some(cpu) = explicit {
        return Ok(cpu);
    }
    if let Some(cpu) = scan_source_cpu(source_text)? {
        return Ok(cpu);
    }
    Err(AsmError::new(
        "CPU が未指定です（--cpu / -m または先頭の .cpu で mn1613 / tms9995 を指定してください）",
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_and_resolve_cpu() {
        let src = ";comment\n.cpu mn1613\nMAIN: h\n";
        assert_eq!(scan_source_cpu(src).expect("scan"), Some(CpuType::Mn1613));
        assert_eq!(
            resolve_cpu_type(None, src).expect("resolve"),
            CpuType::Mn1613
        );
    }
}
